llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -7972,6 +7972,11 @@ Instruction *InstCombinerImpl::visitFCmpInst(FCmpInst &I) {
   Constant *RHSC;
   if (match(Op0, m_Instruction(LHSI)) && match(Op1, m_Constant(RHSC))) {
     switch (LHSI->getOpcode()) {
+    case Instruction::FSub:
+      if (Pred == FCmpInst::FCMP_OGT && match(RHSC, m_PosZeroFP()) &&
```

Comment:
Does it holds for other predicates?
Does it holds when `RHSC` is -0.0?

Reply:
> Does it holds for other predicates?

I think `FCMP_OLT`, `FCMP_ONE` will work. Are there any others that work?

Reply:
the unordered versions of all of them too

Reply:
Yes it should work for -0 as well as +0. Yes it should work for all predicates. But this optimization is not valid if denormals are being flushed - does InstCombine have a way to query that?

Reply:
F.getDenormalType(fltSemantics) == DenormalMode::getIEEE() works, but is pretty ugly

Reply:
I'm encountering issues with some predicates when using alive2 for verification.

Reply:
Can you give an example?

Reply:
https://alive2.llvm.org/ce/z/CgCH-9

Reply:
OK, I think you need an ninf/isKnownNeverInfinity check on the fsub for the unordered cases

Reply:
I see. The tricky case is when both arguments are +inf or both are -inf. The transformation is correct for ordered comparisons that do not include equality (ogt olt one) and unordered comparison that do include equality (ueq uge ule).

Reply:
could I use `m_NonInf` for handling other cases?

Reply:
`m_NonInf` only matches non-inf _constants_. isKnownNeverInfinity is much more powerful because it can tell you when an arbitrary expression is known not to be infinity.

Reply:
```cpp
if (match(RHSC, m_AnyZeroFP()) &&
    match(LHSI, m_FSub(m_Value(X), m_Value(Y))) &&
    isKnownNeverInfinity(X, 0, getSimplifyQuery().getWithInstruction(LHSI)) &&
    isKnownNeverInfinity(Y, 0, getSimplifyQuery().getWithInstruction(LHSI)))
  return new FCmpInst(Pred, X, Y);
```
Is this the correct way to do it? It doesn't seem to be producing the expected output; `isKnownNeverInfinity` produces false for every test.

Reply:
For the tests you have written here, yes it will always fail. You will need to have additional tests covering the no-infinity knowledge.

Also, you're not taking advantage of the fast math flags on the fsub itself by directly looking at the operands. IIRC there's another overload of isKnownNeverInfinity with passed in flags

Reply:
For `ogt olt one ueq uge ule` you do not need to worry about infinity.

For `ugt ult une oeq oge ole` you need to rule out the `+inf - +inf = nan` and `-inf - -inf = nan` cases. To do that, you can check for _any_ of:
- The subtraction has the ninf flag.
- The subtraction has the nnan flag.
- X is known never infinity.
- Y is known never infinity.

Reply:
Or you could just submit a first patch that only handles the easy predicates `ogt olt one ueq uge ule`.

Reply:
As you suggested, I've already incorporated fast flags into the `fsub` operation for some tests. Additionally, I couldn't locate another version of `isKnownNeverInfinity` that accepts flags

Reply:
https://github.com/llvm/llvm-project/blob/6f44bb7717897191be25aa01161831c67cdf5b84/llvm/include/llvm/Analysis/ValueTracking.h#L511 is the helper I was thinking of. It's 1 level below isKnownNeverInfinity, but can be used the same way